radicle-node/src/worker/fetch.rs

```diff
@@ -0,0 +1,330 @@
+mod refspecs;
+pub use refspecs::{AsRefspecs, Rad, Refspec};
+
+mod url;
+use url::LocalUrl;
+
+pub mod error;
+
+use std::collections::BTreeMap;
+use std::ops::Deref;
+
+use nonempty::NonEmpty;
+
+use radicle::crypto::{PublicKey, Unverified, Verified};
+use radicle::prelude::{Doc, Id};
+use radicle::storage::git::Repository;
+use radicle::storage::refs::{SignedRefs, IDENTITY_BRANCH};
+use radicle::storage::{Namespaces, RefUpdate, Remote, RemoteId};
+use radicle::storage::{ReadRepository, ReadStorage, WriteRepository, WriteStorage};
+use radicle::{git, storage, Storage};
+
+/// The initial staging area to perform a fetch from a remote.
+///
+/// The [`StagingRad::refpsecs`] generated are to fetch the `rad/id`
+/// and/or `rad/sigrefs` references from the remote end.
+///
+/// It is then expected to convert this into [`StagingRefs`] using
+/// [`StagingRad::into_staging_refs`] to continue the the rest of the
+/// references.
+pub struct StagingRad<'a> {
+    /// The inner [`Repository`] for staging fetches into.
+    pub(super) repo: StagedRepository,
+    /// The staging [`Storage`] for staging the `repo`.
+    staging: Storage,
+    /// The original [`Storage`] we are finalising changes into.
+    production: &'a Storage,
+    /// The `Namespaces` passed by the fetching caller.
+    namespaces: Namespaces,
+    _tmp: tempfile::TempDir,
+}
+
+/// Indicates whether the innner [`Repository`] is being cloned into
+/// or fetched into.
+pub enum StagedRepository {
+    Cloning(Repository),
+    Fetching(Repository),
+}
+
+impl Deref for StagedRepository {
+    type Target = Repository;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            StagedRepository::Cloning(repo) => repo,
+            StagedRepository::Fetching(repo) => repo,
+        }
+    }
+}
+
+/// The second staging area to perform a fetch from a remote.
+///
+/// The [`StagingRefs::refpsecs`] generated are to fetch any follow-up
+/// references after the fetch on [`StagingRad`]. This may be all the
+/// delegate's references in the case of cloning the new repository,
+/// or it could be fetching the latest updates in the case of fetching
+/// an existing repository.
+///
+/// It is then expected to finalise the process by transferring the
+/// fetched references into the production storage, via
+/// [`StagingRefs::transfer`].
+pub struct StagingRefs<'a> {
+    /// The inner [`Repository`] for staging fetches into.
+    pub(super) repo: StagedRepository,
+    /// The staging [`Storage`] for staging the `repo`.
+    staging: Storage,
+    /// The original [`Storage`] we are finalising changes into.
+    production: &'a Storage,
+    /// The remotes that the fetch is being performed for. These are
+    /// discovered after performing the fetch for [`StagingRad`].
+    remotes: NonEmpty<RemoteId>,
+    _tmp: tempfile::TempDir,
+}
```

Comment:
It seems strange to me that these are two types. I would expect either an `enum` with two variants, or a struct with an `enum` inside of it :thinking: 

Having the two types makes it a bit too "open ended" as to the relationships between them, and how they are to be used.

Reply:
Interesting. I don't think an `enum` would make much sense because then, for example, when we call `transfer` and check that it's in the `rad` fetching phase we have to error.

I thought this was clearer because it splits out the steps into two separate structs -- granted they share a majority of fields. The API _should_ be clear in that `StagingRefs` has no constructor, other than via `StagingRad`, and `StagingRad` is constructed via `new`, and `StagingRefs` has a consuming method `transfer` -- so the path is `StagingRad::new -> StagingRad::into_staging_refs -> StagingRefs::transfer`.